File: q215.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */

// Time Complexity: O(N * log(N))
// Space Complexity: O(1)

var findKthLargest = function (nums, k) {
  const n = nums.length;
  nums.sort((a, b) => a - b);
  return nums[n - k];
};

console.log(findKthLargest(
  [3, 2, 1, 5, 6, 4],
  2,
)); // => 5

console.log(findKthLargest(
  [3, 2, 3, 1, 2, 4, 5, 5, 6],
  4,
)); // => 4

File: q215.java
// O(N lg K) running time + O(K) memory
// Other possibility is to use a min oriented priority queue that will store the K-th largest values. The algorithm iterates over the whole input and maintains the size of priority queue.

class Solution {
  public int findKthLargest(int[] nums, int k) {
    final PriorityQueue < Integer > pq = new PriorityQueue < > ();
    for (int val: nums) {
      pq.offer(val);

      if (pq.size() > k) {
        pq.poll();
      }
    }
    return pq.peek();
  }
}